### PR TITLE
Hybrid float via cap

### PR DESCRIPTION
**Pull request by arichardson**
Extracted from #473

---
This change is [Reviewable](https://reviewable.io/reviews/ctsrd-cheri/llvm-project/488)

### CODE REVIEW

File: llvm/test/CodeGen/RISCV/cheri/double-mem.ll [hunk 1]
```diff
@@ -0,0 +1,75 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
```
Comment (jrtc27) [inline llvm/test/CodeGen/RISCV/cheri/double-mem.ll:1]: This is currently just a copy of float-mem other than the comment on the line below.

File: llvm/test/CodeGen/RISCV/cheri/float-mem.ll [hunk 2]
```diff
@@ -0,0 +1,114 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
```
Comment (jrtc27) [inline llvm/test/CodeGen/RISCV/cheri/float-mem.ll:1]: Use a hard-float ABI for the `+f` tests as I think that'll exercise more of the backend code, whilst also removing some of the `fmv`s here that are about the float calling convention.

File: llvm/test/CodeGen/RISCV/cheri/float-mem.ll [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+; Check that we can load/store single-precision floating point values via a
+; capability pointer without capability mode.
+; RUN: %riscv32_cheri_llc -mattr=-f -verify-machineinstrs < %s \
+; RUN:   | FileCheck -check-prefix=RV32IXCHERI %s
```
Comment (jrtc27) [inline llvm/test/CodeGen/RISCV/cheri/float-mem.ll:5]: I'd like these to be explicit about being hybrid. Generally I've gone with ARCH-ABI for the CHECK lines as that seems to be what RISC-V does, so RV32IXCHERI-ILP32, RV32IFXCHERI-ILP32F etc, and similarly for what are currently PURECAP-*.

File: llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td [hunk 4]
```diff
@@ -1521,6 +1517,12 @@ defm : CheriStPat<store, CFSW, FPR32>;
 
 } // Predicates = [HasCheri, HasStdExtF, IsCapMode]
 
+let Predicates = [HasCheri, HasStdExtF, NotCapMode] in {
+// TODO: We may want to add int<->float patterns for pointers in capability mode.
```
Comment (jrtc27) [inline llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td:1521]: I'm to blame for the bad wording, but:
```suggestion
// TODO: We may want to add int<->float patterns for integer pointers in capability mode.
```

**Review by jrtc27 — CHANGES_REQUESTED** [review]
Some comments on the tests (and a comment on a TableGen comment), but the patterns look fine. I do though wonder if as a longer-term thing we should be thinking about making the legaliser handle this in a target-independent way through TLI hooks (though that can also apply to BuildPairF64/SplitF64 upstream...).

File: llvm/test/CodeGen/RISCV/cheri/float-mem.ll (hunk 3 above)
Comment (arichardson) [reply llvm/test/CodeGen/RISCV/cheri/float-mem.ll:5]: Sounds good, will update tomorrow morning.

File: llvm/test/CodeGen/RISCV/cheri/float-mem.ll [hunk 5]
```diff
@@ -0,0 +1,112 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; Check that we can load/store single-precision floating point values via a
+; capability pointer without capability mode.
```
Comment (jrtc27) [inline llvm/test/CodeGen/RISCV/cheri/float-mem.ll:3]: I'd drop the comment, we're testing both here and the name should make it very obvious what the test is for (plus the IR is trivial).

File: llvm/test/CodeGen/RISCV/cheri/float-mem.ll [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+; RUN: llc -mtriple=riscv32 -target-abi ilp32 -mattr=+xcheri,-cap-mode,-f,-d -verify-machineinstrs < %s \
+; RUN:   | FileCheck -check-prefix=RV32IXCHERI-ILP32 %s
+; RUN: llc -mtriple=riscv32 -target-abi ilp32d -mattr=+xcheri,-cap-mode,+f,+d -verify-machineinstrs < %s \
+; RUN:   | FileCheck -check-prefix=RV32IXCHERI-ILP32D %s
+; RUN: llc -mtriple=riscv64 -target-abi lp64 -mattr=+xcheri,-cap-mode,-f,-d -verify-machineinstrs < %s \
+; RUN:   | FileCheck -check-prefix=RV64IXCHERI-LP64 %s
+; RUN: llc -mtriple=riscv64 -target-abi lp64d -mattr=+xcheri,-cap-mode,+f,+d -verify-machineinstrs < %s \
+; RUN:   | FileCheck -check-prefix=RV64IXCHERI-LP64D %s
+
```
Comment (jrtc27) [inline llvm/test/CodeGen/RISCV/cheri/float-mem.ll:12]: I personally don't like the gap; it's a bit arbitrary whether you group by hybrid/purecap or XLEN first.

**Review by jrtc27 — APPROVED** [review]
A couple of minor comments but otherwise looks good to merge once those have been addressed (which can include convincing me I'm wrong :P).

**Comment by jrtc27** [comment]
(and of course the comments apply to the double test too)

File: llvm/test/CodeGen/RISCV/cheri/float-mem.ll (hunk 5 above)
Comment (arichardson) [reply llvm/test/CodeGen/RISCV/cheri/float-mem.ll:3]: Happy to drop it. I've just got into the habit of adding a short summary of what the test is supposed to do so I don't have to guess or look at the git log when I come back to it at some point in the future.

File: llvm/test/CodeGen/RISCV/cheri/float-mem.ll (hunk 6 above)
Comment (arichardson) [reply llvm/test/CodeGen/RISCV/cheri/float-mem.ll:12]: Will remove, the gap is just there because that's the order in which I added them.

**Pull request merged**